rosidl_runtime_rs/src/string.rs

```diff
@@ -0,0 +1,451 @@
+/// assert!(maybe_str.is_err());
+/// maybe_str = BoundedString::<3>::try_from("ok!");
+/// assert!(maybe_str.is_ok());
+/// let bounded_str = maybe_str.unwrap();
+/// assert_eq!(&bounded_str.to_string(), "ok!");
+/// ```
+#[derive(Clone, Default, Hash, PartialEq, Eq, PartialOrd, Ord)]
+#[repr(transparent)]
+pub struct BoundedString<const N: usize> {
+    inner: String,
+}
+
+/// A zero-terminated string of 16-bit characters with a length limit.
+///
+/// The same as [`WString`], but it cannot be constructed from a string that is too large.
+///
+/// # Example
+///
+/// ```
+/// # use rosidl_runtime_rs::BoundedWString;
+/// # use std::convert::TryFrom;
+/// let mut maybe_wstr = BoundedWString::<3>::try_from("noo!");
+/// assert!(maybe_wstr.is_err());
+/// maybe_wstr = BoundedWString::<3>::try_from("ok!");
+/// assert!(maybe_wstr.is_ok());
+/// let bounded_wstr = maybe_wstr.unwrap();
+/// assert_eq!(&bounded_wstr.to_string(), "ok!");
+/// ```
+#[derive(Clone, Default, Hash, PartialEq, Eq, PartialOrd, Ord)]
+#[repr(transparent)]
+pub struct BoundedWString<const N: usize> {
+    inner: WString,
+}
+
+/// Error type for [`BoundedString::try_from()`] and [`BoundedWString::try_from()`].
+#[derive(Debug)]
+pub struct StringExceedsBoundsError {
+    len: usize,
+    upper_bound: usize,
+}
+
+// ========================= impls for String and WString =========================
+
+// There is a lot of redundancy between String and WString, which this macro aims to reduce.
+macro_rules! string_impl {
+    ($string:ty, $char_type:ty, $string_conversion_func:ident, $init:ident, $fini:ident, $assignn:ident, $sequence_init:ident, $sequence_fini:ident, $sequence_copy:ident) => {
+        #[link(name = "rosidl_runtime_c")]
+        extern "C" {
+            fn $init(s: *mut $string) -> bool;
+            fn $fini(s: *mut $string);
+            fn $assignn(s: *mut $string, value: *const $char_type, n: libc::size_t) -> bool;
+            fn $sequence_init(seq: *mut Sequence<$string>, size: libc::size_t) -> bool;
+            fn $sequence_fini(seq: *mut Sequence<$string>);
+            fn $sequence_copy(
+                in_seq: *const Sequence<$string>,
+                out_seq: *mut Sequence<$string>,
+            ) -> bool;
+        }
+
+        impl Default for $string {
+            fn default() -> Self {
+                let mut msg = Self {
+                    data: std::ptr::null_mut(),
+                    size: 0,
+                    capacity: 0,
+                };
+                // SAFETY: Passing in a zeroed string is safe.
+                if !unsafe { $init(&mut msg as *mut _) } {
+                    panic!("Sinit failed");
+                }
+                msg
+            }
+        }
+
+        impl Clone for $string {
+            fn clone(&self) -> Self {
+                let mut msg = Self::default();
+                // SAFETY: This is doing the same thing as rosidl_runtime_c__String__copy.
```

Comment:
I'm not sure this comment addresses _why_ this is safe. I feel like these safety comments should be written so that someone won't need to dive into an external C library to verify, unless they really wanted to.

Something like: "This is safe because neither `self.data` nor `self.size` are mutated, and `self.data` is guaranteed to be pointing to valid memory by this point, and `self.size` is guaranteed to be correct"?